ddev/ddev
# feat: Allow traefik merged dynamic configuration override with <project>.yaml

## The Issue

Currently the only way to define any Traefik Middleware (or modify anything at all in the Traefik configs) is by removing `#ddev-generated` from the `/pkg/ddevapp/traefik_config_template.yaml` and `~/.ddev/traefik/static_config` which, as described in #6357, invites all sorts of other issues. The solution is to make all `#ddev-generated` files extensible. 

This was recently implemented for the static config via #6317. This PR is for the dynamic config



## How This PR Solves The Issue

This PR adds a generalized merge mechanism to allow extra `/project/.ddev/traefik/dynamic_config.*.yaml` files to be merged into the `<project>.yaml` file, with the last alphanumeric file "winning". This can be used for adding middlewares as well as overriding any field in the dynamic config.

The `dynamic_config.*.yaml` files will also be able to contain {{go templates}} so that middlewares can dynamically use app variables to do things like rewrite app URLs in the response body (useful for Wordpress and Magento), or automatically "namespace" the middleware per active DDEV project.

Most users should never need to touch this - it would be best implemented via addons or other core code. 

## Manual Testing Instructions
* create a ddev project with `ddev config`
* copy/rename `/project/traefik/dynamic_config.middleware.yaml.example` to remove `.example` extension, which will use the fail2ban plugin
* Likewise for ~/.ddev/traefik/static_config.plugins.yaml` to add fail2ban globally
* `ddev start` 
* Check that routers and middleware are loading and working properly at http://localhost:10999/dashboard/#/
* Do the same for a 2nd project and check the traefik dashboard to see that separate middleware has been created per router
* Feel free to try defining other middlewares in both of those files, or changing the router assignments. Make sure to follow the same formatting, particularly with the `{{ .App.Name }}-` prefixes
* Other fields in the dynamic config can be changed as well - e.g. `tls:`


## Automated Testing Overview

A test has been added that injects a middleware definition via an extra config file. 

## Release/Deployment Notes

This should not affect anything else. It adds opt-in functionality, which would be best implemented via addons or other built-in mechanisms.
